Added resolve_customer lambda handler

schaefi opened the pull request:
The resolve_customer handler expects an event providing the x-amzn-marketplace-token. It will then resolve customer information from the metering marketplace as well as the entitlements for this customer's ID and product code. The return type API is documented in the header of the lambda_handler entrypoint

schaefi committed:
Added resolve_customer lambda handler

The resolve_customer handler expects an event providing
the x-amzn-marketplace-token. It will then resolve customer
information from the metering marketplace as well as the
entitlements for this customer's ID and product code.
The input and return API is documented in the header of the
lambda_handler entrypoint and follows the AWS API Gateway
specifications.

schaefi commented:
ok problem with this code is that I implemented it according to the AWS documentation. I was not able to test this code for real because that requires me to have a proper x-amzn-marketplace-token which I don't have and which is the result of a click on the subscription vendor button.

rjschwei commented on aws/resolve_customer/resolve_customer/app.py:
```diff
@@ -0,0 +1,80 @@
+            'CustomerIdentifier': customer.get_id(),
+            'CustomerAWSAccountId': customer.get_account_id(),
+            'ProductCode': customer.get_product_code(),
+            'Entitlements': entitlements.get_entitlements()
```
We will need to keep track of this as a discussion topic with the SCC team as they will receive the information. Please add it to our issues tracker.

schaefi replied:
done

rjschwei commented on aws/resolve_customer/resolve_customer/customer.py:
```diff
@@ -0,0 +1,56 @@
+        self.error = ''
+        if token:
+            try:
+                marketplace = boto3.client('meteringmarketplace')
```
Since we are not handling credentials explicitly we need to have IAM permissions for the lambda function to access the AWS API. Please add this as an item to track in our tasks and issue page.

smarlowucf replied:
I think these would be attached to the lambda function role

schaefi replied:
I added the information to the issue tracker

@smarlowucf regarding access to API's from lambdas. I have done some tests with SQS and I needed to set the following AccessPolicy such that I could access the queue that I created in my account

```
 "Principal": "*",
 "Action": "SQS:*",
```

And I believe this is not the best way to do it. That access thing I leave to the SRE team and my knowledge about access policies and iam roles is very rusty

schaefi replied:
@smarlowucf sorry I deleted a comment from you for which the above comment from me was a reaction to. Your comment appeared twice and I deleted the one from the main page, but that also deleted it here from this conversation. Very weird. Sorry about it

rjschwei commented on aws/resolve_customer/resolve_customer/entitlements.py:
```diff
@@ -0,0 +1,51 @@
+        self.error = ''
+        if customer_id and product_code:
+            try:
+                marketplace = boto3.client('marketplace-entitlement')
```
Another IAM role permission we need.

schaefi replied:
I mentioned both in one issue description, done

schaefi commented:
It's expected that the resolve_customer lambda is triggered by an AWS API Gateway. This information should be added to the code and the code also needs to change because the event structure is not correctly read by the PR for this trigger mechanics. Here is an example event source 

```
[INFO]	2025-01-15T16:44:01.777Z	1c84a83d-f813-4e1a-bacc-c42ab75ebf7d	EVENT: {'version': '2.0', 'routeKey': 'ANY /ms-lambda-from-container', 'rawPath': '/default/ms-lambda-from-container', 'rawQueryString': '', 'headers': {'accept': '*/*', 'content-length': '31', 'content-type': 'application/x-www-form-urlencoded', 'host': 'jfh2r389u9.execute-api.eu-central-1.amazonaws.com', 'user-agent': 'curl/8.0.1', 'x-amzn-trace-id': 'Root=1-6787e5d1-5850b9f64c02c6df0926b405', 'x-forwarded-for': '79.201.150.192', 'x-forwarded-port': '443', 'x-forwarded-proto': 'https'}, 'requestContext': {'accountId': '810320120389', 'apiId': 'jfh2r389u9', 'domainName': 'jfh2r389u9.execute-api.eu-central-1.amazonaws.com', 'domainPrefix': 'jfh2r389u9', 'http': {'method': 'POST', 'path': '/default/ms-lambda-from-container', 'protocol': 'HTTP/1.1', 'sourceIp': '79.201.150.192', 'userAgent': 'curl/8.0.1'}, 'requestId': 'EcDY1h1zFiAEPLQ=', 'routeKey': 'ANY /ms-lambda-from-container', 'stage': 'default', 'time': '15/Jan/2025:16:44:01 +0000', 'timeEpoch': 1736959441730}, 'body': 'eyJuYW1lIjogIk1hcmN1cyIsICJhZ2UiOiAiNTIifQ==', 'isBase64Encoded': True}
```

- [x] code must check body encoding type
- [x] real request event data is below body key, adapt handler method accordingly
- [x] response output must include at least: isBase64Encoded, statusCode and body, see https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-input-format for details

smarlowucf commented on aws/resolve_customer/resolve_customer/app.py:
Does this dir step on the namespace of the above module?

schaefi replied:
There is `aws/resolve_customer/resolve_customer.py` which I plan to install to `/usr/bin` in the spec file. On AWS lambda it can then be registered as

```
CMD override

/usr/bin/resolve_customer.lambda_handler
```

This worked for me and allows standard python packaging. If you have a better way or see issues with the above please don't hesitate to share your ideas. It's also the first time for me that I provide lambda code to the cloud. Thanks

schaefi replied:
@smarlowucf ok you are right, this does not work because lambda doesn't like the sub directories. I'm going to fix this. Just need to read first what's the easiest way

schaefi replied:
finally I managed to get things accepted by lambda. The entry point definition for the AWS lambda runtime interface is now set to 

```
app_resolve_customer.lambda_handler
```

Please note, AWS lambda does not easily allow subdirs to be processed. One would need to change the directory first and I don't wanted to do that. Thus the application entry code is packaged to the root of the system and prefixed with `app_` 

When you start to add your sqs code it would be nice if you can use the same concept. Of course if you know a better way we can also change to it, but we should be consistent in the entry-point handling for all services we offer here.

Thanks much

smarlowucf replied:
I was just playing around with code directly in lambda not in an uploaded container yet. But I remember having trouble with namespaces before in lambda.

smarlowucf replied:
It looks like you were able to get the modules to install as site packages and the imports to work though?

schaefi committed:
Update spec file

Add resolve_customer.py as callable entry point for the lambda
Also fix spec file errors for requirements and a mistake on the
%description field
